test/model/test_lstnet.py

```diff
@@ -0,0 +1,111 @@
+
+# Third-party imports
+import pytest
+import numpy as np
+import pandas as pd
+
+# First-party imports
+from gluonts.dataset.artificial import constant_dataset
+from gluonts.dataset.common import TrainDatasets
+from gluonts.evaluation.backtest import backtest_metrics
+from gluonts.model.lstnet import LSTNetEstimator
+from gluonts.trainer import Trainer
+from gluonts.dataset.multivariate_grouper import MultivariateGrouper
+from gluonts.evaluation import MultivariateEvaluator
+from gluonts.evaluation.backtest import make_evaluation_predictions
+
+
+NUM_SERIES = 10
+NUM_SAMPLES = 5
+
+
+def load_multivariate_constant_dataset():
+    metadata, train_ds, test_ds = constant_dataset()
+    grouper_train = MultivariateGrouper(max_target_dim=NUM_SERIES)
+    grouper_test = MultivariateGrouper(max_target_dim=NUM_SERIES)
+    return TrainDatasets(
+        metadata=metadata,
+        train=grouper_train(train_ds),
+        test=grouper_test(test_ds),
+    )
+
+
+dataset = load_multivariate_constant_dataset()
+freq = dataset.metadata.metadata.freq
+prediction_length = dataset.metadata.prediction_length
+
+
+@pytest.mark.parametrize("skip_size", [1, 2])
+@pytest.mark.parametrize("ar_window", [1, 2])
+@pytest.mark.parametrize(
+    "horizon, prediction_length",
+    [[prediction_length, None], [None, prediction_length]],
+)
+@pytest.mark.parametrize("hybridize", [True, False])
+@pytest.mark.parametrize("dtype", [np.float32, np.float64])
+def test_lstnet(
+    skip_size, ar_window, horizon, prediction_length, hybridize, dtype
+):
+    estimator = LSTNetEstimator(
+        skip_size=skip_size,
+        ar_window=ar_window,
+        num_series=10,
+        channels=6,
+        kernel_size=3,
+        context_length=4,
+        freq=freq,
+        horizon=horizon,
+        prediction_length=prediction_length,
+        trainer=Trainer(
+            epochs=1, batch_size=2, learning_rate=0.01, hybridize=hybridize
+        ),
+        dtype=dtype,
+    )
+
+    predictor = estimator.train(dataset.train)
+    forecast_it, ts_it = make_evaluation_predictions(
+        dataset=dataset.test, predictor=predictor, num_samples=NUM_SAMPLES
+    )
+    forecasts = list(forecast_it)
+    tss = list(ts_it)
+    assert len(forecasts) == len(tss) == len(dataset.test)
+    test_ds = dataset.test.list_data[0]
+    for fct in forecasts:
+        assert fct.freq == freq
+        if estimator.horizon:
+            assert fct.samples.shape == (NUM_SAMPLES, 1, NUM_SERIES)
+        else:
+            assert fct.samples.shape == (
+                NUM_SAMPLES,
+                prediction_length,
+                NUM_SERIES,
+            )
+        assert (
+            fct.start_date
+            == pd.date_range(
+                start=str(test_ds["start"]),
+                periods=test_ds["target"].shape[1],  # number of test periods
+                freq=freq,
+                closed="right",
+            )[-(horizon or prediction_length)]
+        )
```

Comment:
I don't think this is the right assertion in case `horizon` is set: if `horizon` or `prediction_length` are set to `p`, then the forecast contains the following time steps

```
horizon=p           =>               y_{t+p}
prediction_length=p => y_{t+1}, ..., y_{t+p}
```

So, when `horizon` is set, the `fct.start_date` should be `p-1` time steps later, while according to the assertion it's the same in both cases.

However, at the moment there might be no easy way to get the correct `fct.start_date` out of the predictor in that case: I've opened issue #677 in this regard (in particular, the `ForecastGenerator` needs to be modified). I think it's fine for now to have the `prediction_length` case doing the right thing (and the test looks OK for that), and then fix the `horizon` case once #677 is addressed.

Reply:
Right! `start_date` is not available correctly when `horizon` is set.

Reply:
Right! seems the correct `start_date` doesn't seem be available. Though in test case, it's actually `p=2` and the `start_date` is one before the last time.